### ISSUE TITLE
Separate the Conditional sampler for continous and discrete features.

### ISSUE THREAD

**Issue by lionelkusch** [mind-inria/hidimstat#454]
Following the conversation in the issue #445 and the implementation of conditional sampler: https://github.com/mind-inria/hidimstat/blob/a8e1c7322a5783c741c030f456e5f27cfa37e5aa/src/hidimstat/conditional_sampling.py#L42-L46

It seems that the conditional sampler can be reused between methods and there are not necessarily the same for continuous and discrete features.

Do you think that separate this creates a class of conditional samplers by type and let the methods choice the right conditional sampler to use be a good design to choose?
The second possibility is to have conditional samplers, which should always handle these 2 types of features.

**Comment by bthirion** [comment]
IIUC, you mean that we don't want to support multiple conditional samplers in the library, which makes sense.
Unless I'm wrong , only knockoff (DCRT ?) and CPI require such a sampler, right ? 
I think it makes sense to factor it out.

**Comment by bthirion** [comment]
Beware the YAGNIs. Most of them are brittle if not broken.

**Comment by lionelkusch** [comment]
When you mean "brittle", do you talk about the implementation or about the theory?

**Comment by bthirion** [comment]
Theory tells you that complex samplers (normalizing flows etc.) trained on limited data may not work well: indeed they don't. 
I'm not criticizing a particular implementation.

**Comment by bthirion** [comment]
I don't want to have YAGNIs: if somebody **needs** normalizing flows (and can explain why), we can add them, but we should not do it just because there is an existing implementation.
I'm not actually sure we need a mixture of Gaussians. 
For categorical variables, we only need a multilogistic model I think.
Did you have anything else in mind ?

**Comment by lionelkusch** [comment]
For categorical variables, I have more a naive Bayes mode in mind but it seems that scikitlearn has already a model for it ([see](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.CategoricalNB.html#sklearn.naive_bayes.CategoricalNB)).
 
I am perhaps wrong but CRT is based on Gaussian mixture distribution. We can perhaps use the Gaussian mixture proposed by [scikitlearn](https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html). I can add adapter to the scikitlearn mixture model in hidimstat.
Flippy proposes a NetworkMixtureGaussian but I think that this case is too specific.

To come back to the original question, Do you think that separate the sampler to categorise and continuous variables and let the method decide what to use for each feature is better or not?

**Comment by bthirion** [comment]
* I think that it is OK that the user provides the information whether he/she wants the variables to be handled as categorical or continuous.
* if the information is not provided, then the decision could be made by the estimator. 
* When we sample, we should start with the continuous variables, then sample the categorical conditional to the continuous ones.
* sklearn's NB is for cases where no continuous variable is available. I think that it is not useful for us (or only in a corner case that we can ignore in a first attempt)
* CRT  does not need GM IIRC
Does that make sense ?

**Comment by bthirion** [comment]
We should consider whether there needs to be some common refactoring among knockoffs, CRT, dCRT and CPI, which would justify having a class.

**Comment by jpaillard** [comment]
As you say, I initially implemented it in one class to handle the data type assignment inside the sampler instead of the parent class. 
I can see the point of separating the sampling into 2 classes, as the process is quite different for continuous/discrete. However, I would keep a meta-class that handles data type and sampler assignment.

**Comment by lionelkusch** [comment]
Why do you want to have a meta-class for handling data type and sampler assignments?
For me, a function in this case seems more appropriate.

**Comment by jpaillard** [comment]
Function would be perfectly fine. I just meant not to make the code of classes like CFI, heavier with this.
